### PR TITLE
Implement v3 api

### PR DESCRIPTION
**Pull request by ypaq**
Logplex's v2 API leaves many opportunities for race conditions with other Heroku internal services. The proposed v3 API makes channel creation idempotent by turning channel IDs to strings and leaving the generation of channel IDs to the caller. 

The API is supposed to be as restful as possible. The [schema](https://github.com/heroku/logplex/blob/73d4ef7ef142019dcb54b0a07cc2000f5ffdb4a1/doc/schema/schema.md) contains a detailed descriptions which endpoints are covered by the v3 API. All v2 API calls to the channels, drains, and tokens endpoint are deprecated. The session endpoints remain in place for now.

### CODE REVIEW

File: src/logplex_api_v3_channels.erl [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+%% @private
+from_json(Req, #state{ channel_id = ChannelId,
+                       tokens     = OldTokens,
+                       drains     = Drains } = State) ->
+    {ok, Body, Req1} = cowboy_req:body(Req),
+    try jsx:decode(Body, [return_maps]) of
+        Map ->
+            case validate_payload(Map) of
+                {ok, NewTokenNames} ->
+                    ok = create_or_update_channel(ChannelId),
```
Comment (voidlock) [inline src/logplex_api_v3_channels.erl]: This triggers a `logplex_channel:store` to happen, which in turn will overwrite any data stored in redis. Won't this mean that any stored flags like `no_redis` will be obliterated by this operation?

File: src/logplex_api_v3_drains.erl [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            %% channel was not found
+            {ok, Req1} = cowboy_req:reply(404, Req),
+            {halt, Req1, State}
+    end;
+resource_exists(Req, #state{ channel_id = ChannelId,
+                             drain_id   = DrainId } = State) ->
+    Timeout = logplex_app:config(default_redis_poll_ms, 2000),
+    case logplex_channel:poll(ChannelId, Timeout) of
+        Channel when is_record(Channel, channel) ->
+            case logplex_drain:poll(DrainId) of
```
Comment (voidlock) [inline src/logplex_api_v3_drains.erl]: Same comment as before, except this time we could hang out waiting for worst case 4 seconds.

File: src/logplex_firehose.erl [hunk 3]
```diff
@@ -40,7 +40,7 @@
 
 -record(shard_pool, {key=master_shard,
                      size=0 :: integer(),
-                     pool={} :: tuple()}).
+                     pool=[] :: list()}).
```
Comment (voidlock) [inline src/logplex_firehose.erl]: Why did this change?

File: src/logplex_api_v3_channels.erl [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+%% @private
+is_authorized(Req, State) ->
+    logplex_api_v3:is_authorized(Req, State).
+
+%% @private
+resource_exists(Req, #state{ channel_id = ChannelId } = State) ->
+    %% we need to poll here to wait for nsync replication to catch up
+    Timeout = logplex_app:config(default_redis_poll_ms, 2000),
+    case logplex_channel:poll(ChannelId, Timeout) of
```
Comment (ypaq) [inline src/logplex_api_v3_channels.erl]: That is exactly what it means.

File: src/logplex_api_v3_channels.erl (hunk 1 above)
Comment (ypaq) [inline src/logplex_api_v3_channels.erl]: Correct, I believe v2 has the same behavior.

File: src/logplex_api_v3_drains.erl (hunk 2 above)
Comment (ypaq) [inline src/logplex_api_v3_drains.erl]: Exactly.

File: src/logplex_firehose.erl (hunk 3 above)
Comment (ypaq) [inline src/logplex_firehose.erl]: Dialyzer type error fix.

**Pull request merged**